test/integration/tcp_proxy_odcds_integration_test.cc

```diff
@@ -0,0 +1,304 @@
+      auto tcp_proxy_config =
+          MessageUtil::anyConvert<envoy::extensions::filters::network::tcp_proxy::v3::TcpProxy>(
+              *config_blob);
+      tcp_proxy_config.set_cluster("new_cluster");
+      tcp_proxy_config.mutable_on_demand()->CopyFrom(
+          TestUtility::parseYaml<
+              envoy::extensions::filters::network::tcp_proxy::v3::TcpProxy_OnDemand>(R"EOF(
+          odcds_config:
+            resource_api_version: V3
+            api_config_source:
+              api_type: DELTA_GRPC
+              transport_api_version: V3
+              grpc_services:
+                envoy_grpc:
+                  cluster_name: cluster_0
+      )EOF"));
+      config_blob->PackFrom(tcp_proxy_config);
+    });
+
+    // The test envoy uses static listener and cluster for xDS.
+    // The test framework does not update the above static resources.
+    // Another upstream will be serving delta CDS requests and the response is explicitly
+    // generated by test case rather than the integration test framework.
+    use_lds_ = false;
+    enableHalfClose(true);
+  }
+  void TearDown() override { cleanUpXdsConnection(); }
+
+  void initialize() override {
+    // The first upstream serves the ADS request including the future on demand CDS request.
+    setUpstreamCount(1);
+    setUpstreamProtocol(FakeHttpConnection::Type::HTTP2);
+
+    BaseIntegrationTest::initialize();
+
+    // HTTP protocol version is not used because tcp stream is expected.
+    addFakeUpstream(FakeHttpConnection::Type::HTTP2);
+
+    new_cluster_ = ConfigHelper::buildStaticCluster(
+        "new_cluster", fake_upstreams_.back()->localAddress()->ip()->port(),
+        Network::Test::getLoopbackAddressString(ipVersion()));
+
+    test_server_->waitUntilListenersReady();
+    registerTestServerPorts({"tcp_proxy"});
+  }
+
+  // The on demand CDS stream. The requested cluster config is delivered to Envoy in this stream.
+  FakeStreamPtr odcds_stream_;
+  // The prepared cluster config response.
+  envoy::config::cluster::v3::Cluster new_cluster_;
+  std::vector<FakeRawConnectionPtr> fake_upstream_connections_;
+  std::vector<IntegrationTcpClientPtr> tcp_clients_;
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersionsClientType, TcpProxyOdcdsIntegrationTest,
+                         GRPC_CLIENT_INTEGRATION_PARAMS);
+
+TEST_P(TcpProxyOdcdsIntegrationTest, SingleTcpClient) {
+  initialize();
+
+  // Establish a tcp request to the Envoy.
+  IntegrationTcpClientPtr tcp_client = makeTcpConnection(lookupPort("tcp_proxy"));
+
+  // The delta CDS stream is established.
+  auto result = fake_upstreams_.front()->waitForHttpConnection(*dispatcher_, xds_connection_);
+  RELEASE_ASSERT(result, result.message());
+  result = xds_connection_->waitForNewStream(*dispatcher_, odcds_stream_);
+  RELEASE_ASSERT(result, result.message());
+  odcds_stream_->startGrpcStream();
+
+  // Verify the delta CDS request and respond with the prepared `new_cluster`.
```

Comment:
Can also wait for `tcp.tcp_stats.on_demand_cluster_attempt` to be 1 to ensure the CDS request received using ODCDS.